File: frontend/src/tools/registry.ts
/**
 * Learning Tool Registry
 *
 * Central registry for all learning tools. Tools register themselves
 * and can be looked up by ID for dynamic rendering.
 */

import { LearningTool, ToolRegistry } from './types';

/**
 * Internal storage for registered tools
 */
const tools = new Map<string, LearningTool>();

/**
 * Register a learning tool
 *
 * @param tool - The tool to register
 * @throws Error if tool with same ID already exists
 */
export function registerTool<TData>(tool: LearningTool<TData>): void {
  if (tools.has(tool.metadata.id)) {
    console.warn(
      `Tool "${tool.metadata.id}" is already registered. Skipping duplicate registration.`
    );
    return;
  }
  tools.set(tool.metadata.id, tool as LearningTool);
}

/**
 * Get a tool by its ID
 *
 * @param id - Tool identifier (e.g., 'quiz', 'lesson')
 * @returns The tool or undefined if not found
 */
export function getTool(id: string): LearningTool | undefined {
  return tools.get(id);
}

/**
 * Get all registered tools
 *
 * @returns Array of all tools
 */
export function getAllTools(): LearningTool[] {
  return Array.from(tools.values());
}

/**
 * Check if a tool is registered
 *
 * @param id - Tool identifier
 * @returns true if tool exists
 */
export function hasTool(id: string): boolean {
  return tools.has(id);
}

/**
 * Get tool descriptions for LLM prompts
 *
 * Returns a formatted string describing all available tools
 * for use in the tool selector prompt.
 *
 * @returns Formatted tool descriptions
 */
export function getToolDescriptions(): string {
  const descriptions: string[] = [];

  for (const tool of tools.values()) {
    const { id, name, icon, description, targetMinutes } = tool.metadata;
    descriptions.push(
      `- **${id}** (${icon} ${name}): ${description} [${targetMinutes} min]`
    );
  }

  return descriptions.join('\n');
}

/**
 * Get tool schemas for backend validation
 *
 * Returns all tool schemas as a map for backend use.
 *
 * @returns Map of tool ID to JSON schema
 */
export function getToolSchemas(): Record<string, object> {
  const schemas: Record<string, object> = {};

  for (const tool of tools.values()) {
    schemas[tool.metadata.id] = tool.schema;
  }

  return schemas;
}

/**
 * Clear all registered tools (useful for testing)
 */
export function clearRegistry(): void {
  tools.clear();
}

/**
 * Registry object for dependency injection
 */
export const toolRegistry: ToolRegistry = {
  register: registerTool,
  get: getTool,
  getAll: getAllTools,
  has: hasTool,
  getToolDescriptions,
};

export default toolRegistry;

File: backend/internal/tools/registry.go
package tools

import (
	"fmt"
	"strings"
	"sync"
)

// Registry holds all registered learning tools.
type Registry struct {
	mu    sync.RWMutex
	tools map[string]LearningTool
}

// Global registry instance
var globalRegistry = &Registry{
	tools: make(map[string]LearningTool),
}

// Register adds a tool to the global registry.
func Register(tool LearningTool) {
	globalRegistry.mu.Lock()
	defer globalRegistry.mu.Unlock()

	id := tool.Metadata().ID
	if _, exists := globalRegistry.tools[id]; exists {
		// Already registered, skip
		return
	}
	globalRegistry.tools[id] = tool
}

// Get retrieves a tool by ID from the global registry.
func Get(id string) (LearningTool, bool) {
	globalRegistry.mu.RLock()
	defer globalRegistry.mu.RUnlock()

	tool, exists := globalRegistry.tools[id]
	return tool, exists
}

// GetAll returns all registered tools.
func GetAll() []LearningTool {
	globalRegistry.mu.RLock()
	defer globalRegistry.mu.RUnlock()

	result := make([]LearningTool, 0, len(globalRegistry.tools))
	for _, tool := range globalRegistry.tools {
		result = append(result, tool)
	}
	return result
}

// Has checks if a tool is registered.
func Has(id string) bool {
	globalRegistry.mu.RLock()
	defer globalRegistry.mu.RUnlock()

	_, exists := globalRegistry.tools[id]
	return exists
}

// GetToolDescriptions returns a formatted string describing all tools
// for use in LLM prompts.
func GetToolDescriptions() string {
	globalRegistry.mu.RLock()
	defer globalRegistry.mu.RUnlock()

	var descriptions []string
	for _, tool := range globalRegistry.tools {
		meta := tool.Metadata()
		desc := fmt.Sprintf("- **%s** (%s %s): %s [%d min]",
			meta.ID, meta.Icon, meta.Name, meta.Description, meta.TargetMinutes)
		descriptions = append(descriptions, desc)
	}
	return strings.Join(descriptions, "\n")
}

// GetToolIDs returns all registered tool IDs.
func GetToolIDs() []string {
	globalRegistry.mu.RLock()
	defer globalRegistry.mu.RUnlock()

	ids := make([]string, 0, len(globalRegistry.tools))
	for id := range globalRegistry.tools {
		ids = append(ids, id)
	}
	return ids
}

// GetSchemas returns all tool schemas as a map.
func GetSchemas() map[string]JSONSchema {
	globalRegistry.mu.RLock()
	defer globalRegistry.mu.RUnlock()

	schemas := make(map[string]JSONSchema)
	for id, tool := range globalRegistry.tools {
		schemas[id] = tool.Schema()
	}
	return schemas
}

// Clear removes all tools from the registry (for testing).
func Clear() {
	globalRegistry.mu.Lock()
	defer globalRegistry.mu.Unlock()

	globalRegistry.tools = make(map[string]LearningTool)
}
